### ISSUE TITLE
TOC does not work when importing one MDX into another

### ISSUE THREAD

**Issue by oriooctopus** [facebook/docusaurus#3915]
## 🐛 Bug Report

On our site we have a couple of duplicate doc pages. To avoid repeating content we have one doc as the source of truth then import the content like so:

```
import Content from './doc1.md';

<Content />
```

The issue is, the table of contents is not picked up on the pages that import the content. Please let me know if there is a better way to handle duplicate pages and if this usage is just conceptually wrong

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## To Reproduce

npx @docusaurus/init@latest init my-website classic

in doc2:
```
import Content from './doc1.md';

<Content />
```

It's worth noting that this also happens on my 6 month old version, so I don't think that it's due to anything recent

**Comment by slorber** [comment]
This usage should be supported, but it's a current limitation of the existing system.

Unfortunately, I'm not sure how to implement this technically, and still busy on i18n.

In the meantime I'd recommend to only put raw text in mdx  and duplicate the headings. That's not ideal but I don't have any other solution for now 😓

**Comment by oriooctopus** [comment]
How difficult would it be to fix? I could possibly put up a PR

**Comment by slorber** [comment]
@oriooctopus unfortunately I suspect this to be quite complex to implement, and can't even give you much help to do so because I'm not sure how this should be done.

You can look at the code in this folder: packages/docusaurus-mdx-loader/src/remark/toc
If you are comfortable with ASTs and visitors you can give it a try

**Comment by jknoxville** [comment]
@slorber I think we can address this by getting docusaurus-mdx-loader to follow jsx tags into their imported sources when traversing the AST.

I did something similar to this before, using the `fs` api to read the file and getting remark to parse this in-place of the jsx node (so you can continue to recurse into the imported file and visit its nodes).

The only downside in my eyes is that we'd be using the node fs api, which means this package wouldn't work at runtime in the browser anymore (if it does now).

Do you think this would be an acceptable change?

**Comment by slorber** [comment]
Yes @jknoxville this is what I have in mind: a recursive visitor that is able to "flatten" the TOC structure of multiple MDX docs (and preserve the correct TOC ordering).

This code already runs in node (it's a Webpack loader)

**Comment by oriooctopus** [comment]
@jknoxville do you still have any interest in fixing this? If not, maybe we could sync on what you had in mind and I might be able to implement it?

**Comment by jknoxville** [comment]
@oriooctopus I'm interested, but it's not top of my queue right now, so feel free to take a stab!

Here's a prototype I made a while ago, for a remark plugin that effectively re-implements mdx transclusions by parsing the target of any imported markdown files, and then replacing any JSX nodes that use it, with the nodes of the target AST.
https://gist.github.com/jknoxville/8ed4cb8bcef348362a99dd70bb8006ae

It's scrappy, and not production-ready (currently uses eval for one thing - I'd recommend using babel to parse the imports instead), but it works. Maybe there's a more elegant way to do it.

What I imagine, is modifying [search.js](https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-mdx-loader/src/remark/toc/search.js#L66) to do what that prototype does before it starts visiting headings to assemble its TOC.

I haven't tried this, but don't see why it wouldn't work.

**Comment by slorber** [comment]
if that helps I remembered that Gatsby has a way to extract all imports/exports from MDX.

https://github.com/gatsbyjs/gatsby/blob/2e42197025e2e1bac06c721c3cc44135bf8ef526/packages/gatsby-plugin-mdx/utils/gen-mdx.js#L199

**Comment by Laptopmini** [comment]
Can the table of content be overwritten using a React component? 

I'm experiencing a similar issue where content generated via a React component is not parsed by the table of contents (for obvious reasons), which got me thinking about if there was a way to provide a custom table of content for certain pages.

**Comment by slorber** [comment]
@Laptopmini, I think extracting headings from React components is an edge case, more complicated and expensive (as it would require rendering the comp to see what's the output), so it's not likely the first thing we could build.

We could allow you to export a custom toc in the MDX document, or modify the default one we computed for you?

```mdx
---
id: myDoc
--- 

export function toc(originalToc) {
  return [...originalToc,{value: 'My React heading',id: 'my react anchor id',children: []}]
}

# Title

## Subtitle

Lorem Ipsum
```

Does it look like a good workaround for your usecase?

**Comment by Laptopmini** [comment]
I found a solution to my scenario, where I use a React portal to inject a custom table of contents, reusing the same structure and classNames as the current theme. The className usage is hacky and subject to breaking but it works well and should serve fine for the time being.

```js
const TableOfContents = ({ items }) => {
  if (!items.length) return null;
  return useMemo(() => {
    const [parent] = document.getElementsByClassName(
      'tableOfContents_node_modules-@docusaurus-theme-classic-lib-next-theme-TOC-'
    );
    return ReactDOM.createPortal(
      <ul className="table-of-contents table-of-contents__left-border">
        {items.map(({ id }) => {
          return (
            <li key={id}>
              <a
                href={`#${id}`}
                className="table-of-contents__link table-of-contents__link--active"
              >
                {id}
              </a>
            </li>
          );
        })}
      </ul>,
      parent
    );
  }, [items]);
};
```

```js
<TableOfContents items={[{ id: "example1" }, { id: "example2" }]} />
```

**Comment by sweeneyskirt-sl** [comment]
Hi @Laptopmini  -- I would love to try to replicate your workaround to this issue, but I am not quite as savvy with respect to coding the backend. Could you provide some more specifics about where, exactly, you added the react component above, or provide a link to your repo where you have this working so I can look at it? Thanks!

**Comment by Laptopmini** [comment]
@sweeneyskirt-sl It seems like docusaurus allocates a column in the page for a possible Table Of Contents, regardless of if its created or not by the library. You can verify this by checking for the presence of an element with the class `tableOfContents_node_modules-@docusaurus-theme-classic-lib-next-theme-TOC-` in the right hand side of the page. It is within the rightmost `col` in the DOM, the class name is subject to change depending on the theme you are using it seems.

If you are using the same theme, or adapting the definition of `TableOfContents` I wrote to target the right class, you can then use `<TableOfContents items={[{ id: "example1" }, { id: "example2" }]} />` in a `.mdx` file and it should inject a custom table of contents into the page based on the `items` you pass it.

**Comment by slorber** [comment]
@Laptopmini what I understand is that you want to create a TOC dynamically from data fetched from an API.

This is unrelated to the current issue which is about importing one MDX doc inside another.

Please open another issue if you want to discuss dynamically generated TOC, and let's keep this issue focused on its initial problem. I believe you could just swizzle the TOC component and add your stateful logic here.

**Comment by sweeneyskirt-sl** [comment]
Hi @slorber -- Thanks for your reply. So I have a file that defines a number of configuration variables that are common to several environments, so I have a single doc that I am importing into the other docs, but I need each of the variables to be included in the floating TOC, which they are not, as clearly confirmed by this issue. I see from your earlier response that fixing this issue is not a priority for docusauraus right now -- do you know if it will be at any point in the near roadmap? Are there any viable workarounds? Manually adding a header for each variable is not viable for my content. It would defeat the purpose of the transclusion. Thank you for any advice!

**Comment by khushal87** [comment]
Any solution for this? @slorber @oriooctopus

**Comment by sweeneyskirt-sl** [comment]
I have not heard of any updates to this issue yet, @khushal87. I did throw in a feature request at https://docusaurus.canny.io/feature-requests/p/include-transclusion-content-headings-in-toc so please feel free to vote it up so we can get this on the roadmap at least! :)

**Comment by khushal87** [comment]
Done @sweeneyskirt-sl :crossed_fingers:

**Comment by slorber** [comment]
Some related notes:

- We added better support for the _ prefix convention so that doc partials are not creating additional routes: https://github.com/facebook/docusaurus/pull/5173

- Instead of using MDX imports (which don't work with the TOC), we may try to explore a way to include one file's content into another. MDX will process the md content as a whole (1 final React component instead of multiples) and be able to compute the appropriate TOC. This solution has been mentioned here already: https://github.com/facebook/docusaurus/issues/599#issuecomment-605847247 Please let me know if that could work for your usecase?

**Comment by slorber** [comment]
A temporary workaround could be to export your own TOC manually, eventually re-using the TOC exported by the MDX partials you used. 

I've used this trick on our changelog page, that imports the changelog from the root of the repo: https://github.com/facebook/docusaurus/pull/5331

This can work well particularly if your existing doc is just an "index" for many imported docs and does not declare its own headings:

```mdx
import Chapter1, {toc as Chapter1TOC} from "_chapter1.md"
import Chapter2, {toc as Chapter2TOC} from "_chapter2.md"

<Chapter1 />

<Chapter2 />

export const toc = [...Chapter1TOC, ...Chapter2TOC];
```

This becomes less usable once you add headings between the 2 imported chapters

**Comment by Josh-Cena** [comment]
We don't make any assumptions. You insert an H1 heading, and it is an H1 heading. We call these imported files "partials" because, well, they are just "partials".

**Comment by Josh-Cena** [comment]
@ltribolet If the page you are importing the partial to already has a TOC, then no, you would probably need something discussed in #6201 to do it.

**Comment by Yukiniro** [comment]
> A temporary workaround could be to export your own TOC manually, eventually re-using the TOC exported by the MDX partials you used.
> 
> I've used this trick on our changelog page, that imports the changelog from the root of the repo: #5331
> 
> This can work well particularly if your existing doc is just an "index" for many imported docs and does not declare its own headings:
> 
> ```gfm
> import Chapter1, {toc as Chapter1TOC} from "_chapter1.md"
> import Chapter2, {toc as Chapter2TOC} from "_chapter2.md"
> 
> <Chapter1 />
> 
> <Chapter2 />
> 
> export const toc = [...Chapter1TOC, ...Chapter2TOC];
> ```
> 
> This becomes less usable once you add headings between the 2 imported chapters

It is working.

**Comment by Josh-Cena** [comment]
After #6729, we are in a better position of fixing this problem. If anyone wants to take a look at this issue, here's how we should solve this:

1. The TOC remark plugin should recognize import statements that import a Markdown document, and add a named import called `toc` if it's not already imported. e.g.

```diff
  // Non-MD files are ignored
  import Partial1 from "./Partial1.jsx";
  // Imports already containing a `toc` name are kept as-is
  import Partial2, { toc as Partial2TOC } from "./Partial2.mdx";
  import Partial3, { toc } from "./Partial3.md";
  // Automatically append a `toc` import. The name must be unique to avoid conflicts.
  // I think we can simply use "__toc" + name of the partial component
- import Partial4 from "./Partial4.mdx";
+ import Partial4, { toc as __tocPartial4 } from "./Partial4.mdx";
- import Partial5, { frontMatter } from "./Partial5.md";
+ import Partial5, { frontMatter, toc as __tocPartial5 } from "./Partial5.md";
```

Simultaneously, this traversal step needs to record the names of all Markdown components in scope.

2. The TOC collection part (the visitor), in addition to visiting heading nodes, should also keep track of where these Markdown partials are rendered. When it encounters `<Partial1 />` where `Partial1` is a Markdown component (known from the previous step), it inserts a dummy placeholder in the `toc` list telling the codegen step that "here's we should insert the TOC from the partial".

In order to know the name of the component being used, we need to parse the JSX expression with, say, Babel, since a `jsx` node has its content perfectly preserved. This should be fine, since JSX in Markdown is usually lightweight so performance is not sacrificed much. We can inspect using SWC later. The TOC algorithm already uses Babel to parse export statements and find existing exports called `toc`.

Note that this may not be as trivial as it seems. Consider the following case:

```jsx
<div>

Some content

</div>
```

Instead of seeing one JSX, the visitor will see three nodes, with the opening tag and the closing tag being different nodes. If we naively pass `<div>` to Babel, the parser will fail. We will need to add a closing tag to let it parse gracefully.

Alternatively, we can be bold and assume that all Markdown partials are their own nodes (i.e. on their own paragraphs, or at least not wrapped by other JSX nodes like `<div><Partial1 /></div>`). In this case, we may get away with simple regex parsing.

3. Now that we have collected the TOC list, which looks something like:

```ts
const toc = [
  {
    value: 'Thanos',
    id: 'thanos',
    level: 2,
  },
  {
    value: 'Tony Stark',
    id: 'tony-stark',
    level: 2,
  },
  {
    name: 'Partial2TOC',
  },
  {
    value: 'Avengers',
    id: 'avengers',
    level: 3,
  },
  {
    name: '__tocPartial4',
  },
];
```

The last step is relatively easy. During codegen, we transform all exotic TOC nodes to a spread:

```ts
export const toc = [
  {
    value: 'Thanos',
    id: 'thanos',
    level: 2,
  },
  {
    value: 'Tony Stark',
    id: 'tony-stark',
    level: 2,
  },
  ...Partial2TOC,
  {
    value: 'Avengers',
    id: 'avengers',
    level: 3,
  },
  ...__tocPartial4,
];
```

And voila! There we get a fully functional TOC. Note the merits of this approach: we never actually traverse the imported Markdown partial. The exotic TOCs remain as opaque data structures.

The main technical difficulties lie in parsing the import statements, adding extra import specifiers, and parsing the JSX nodes. Community contributions welcome.

**Comment by o-v** [comment]
> `import Content from './doc1.md';`

Correct me if I'm wrong, the document gets imported as `MDXComponent` (with headings already rendered as `h2`s so won't have TOCs generated for?)

I suspect [the eventual upgrade to mdxjs2](https://github.com/facebook/docusaurus/issues/4029) will solve this because of [how markdown will be treated differently within JSX blocks](https://mdxjs.com/blog/v2/#improvements-to-the-mdx-format).

**Comment by Josh-Cena** [comment]
No. The reason why there's no TOC is because TOC generation is a Remark plugin that statically analyzes the Markdown document instead of actually running it and collecting h2 headings present in the actual generated JSX. So my proposed solution above involves collecting the incoming `toc` variable of `./doc1.md`, and inserting it into the current file's TOC.

**Comment by Mr-spShuvo** [comment]
@Josh-Cena I understand. TOC of content is curial part of Documentation website. And MDX is indented to make multiple files. I wonder someone raised this issue long ago, yet there is no progress so far. I'm not experienced enough to contribute right now. But I believe this deserves more priority.

**Comment by slorber** [comment]
Note: the readingTime blog feature is also affected by this MDX imports problem.

---

Wondering, how many of you have been successful using a remark plugin to inline one MD doc into another? Have you published your solution?


Looks like this has been used successfully by at least 2 persons:
- https://github.com/facebook/docusaurus/issues/599#issuecomment-605847247
- https://github.com/facebook/docusaurus/issues/3915#issuecomment-763574482

**Comment by breezyfasano** [comment]
Hi there! I actually worked on this for our site and think I may have a potential solution.
View my PR to our public docs site here:
https://github.com/dbt-labs/docs.getdbt.com/pull/2395

Is there anyone that would want to pair on this and see if I can contribute this?

**Comment by B2o5T** [comment]
For anyone waiting to fix this bug, I am happy to say that today I fixed this confusing behaviour in [Nextra@2.12.0](https://github.com/shuding/nextra/releases/tag/nextra%402.12.0) via the custom remark plugin written by me.

After receiving feedback from my community I want to contribute to Docusaurus too via a generic remark plugin that we could use in both frameworks Docusaurus & Nextra.

Stay tuned!

Here you can see a test with snapshot result

https://github.com/shuding/nextra/blob/main/packages/nextra/__test__/compile.test.ts#L83-L123

**Comment by rajkaramchedu** [comment]
> For anyone waiting to fix this bug, I am happy to say that today I fixed this confusing behaviour in [Nextra@2.12.0](https://github.com/shuding/nextra/releases/tag/nextra%402.12.0) via the custom remark plugin written by me.
> 
> After receiving feedback from my community I want to contribute to Docusaurus too via a generic remark plugin that we could use in both frameworks Docusaurus & Nextra.
> 
> Stay tuned!
> 
> Here you can see a test with snapshot result
> 
> https://github.com/shuding/nextra/blob/main/packages/nextra/__test__/compile.test.ts#L83-L123

Wonderful. I needed the target file combining the partials to show the right sidebar. Because of this bug, I have decided to take a somewhat circuitous route of adding the `##` headers in the target file and then including the partial below it, for several such subsections. Your fix, when it arrives, will be godsend. Thanks!

**Comment by juniorgasparotto** [comment]
Any news on this subject?

**Comment by albertothedev** [comment]
> Any news on this subject?

Seems like we are waiting for @dimaMachina to create a remark plugin with his fix.

**Comment by dimaMachina** [comment]
> > Any news on this subject?
> 
> Seems like we are waiting for @dimaMachina to create a remark plugin with his fix.

Yes! Sorry, currently I am busy with Nextra 3 release and future speak in Amsterdam on 8 November, I hope to do it next month ✌️

**Comment by ohsayan** [comment]
I think we just ran into this as well. We basically wanted to have a component that renders a small label next to a heading to note if the thing being discussed is a beta item or not. While the component works flawlessly, the TOC does not render the heading. As a workaround, for now is there a way to manually create a TOC for a given MD doc?

**Comment by rajkaramchedu** [comment]
> I think we just ran into this as well. We basically wanted to have a component that renders a small label next to a heading to note if the thing being discussed is a beta item or not. While the component works flawlessly, the TOC does not render the heading. As a workaround, for now is there a way to manually create a TOC for a given MD doc?

This is how I did it. Quite sure there are better ways. I split the Markdown document I wanted to import into multiple MDX files. Each MDX file contains only the content from a specific section of the Markdown doc. So in the below, my MD doc had sections like Description, Parameters and Returns in one doc. After splitting it and saving into multiple files, each beginning with the underscore character, I imported them into another MDX doc like this. This way, Docusaurus displayed all the section headers, like it would for any normal MD doc. 

```
import Description from "/path/to/_partials/_mydoc-description.mdx";

<Description />

## Parameters

import Params from "/path/to/_partials/_mydoc_-parameters.mdx";

<Params />

## Returns

import Returns from "/path/to/_partials/_mydoc_-returns.mdx";

<Returns />
```

**Comment by robodl** [comment]
We've done it like this

```
import Introduction, {
    toc as IntroductionToc,
} from "../shared/introduction.mdx";
import IntroductionContent, {toc as IntroductionContentToc} from "./.introduction_content.mdx";

export const toc = [...IntroductionContentToc, ...IntroductionToc]

<IntroductionContent />

<Introduction />
```

But then you can't mix it with other headers, like this

```
...

export const toc = [...IntroductionContentToc, ...IntroductionToc]

<IntroductionContent />

## Other header <---- WONT WORK PROPERLY

<Introduction />
```

**Comment by anatolykopyl** [comment]
I submitted a PR, that I'll be willing to work on, but in the meantime, until it is accepted, I made a quick drop in replacement remark plugin that fixes this behaviour in Docusaurus v2: [`@akopyl/docusaurus-toc-patcher`](https://github.com/anatolykopyl/docusaurus-toc-patcher).

It implements the logic from https://github.com/shuding/nextra/pull/2199 in Docusaurus v2.

**Comment by slorber** [comment]
This has been fixed (main/canary) as part of https://github.com/facebook/docusaurus/pull/9684, and will be released in Docusaurus v3.2

Thanks for the inspiration @dimaMachina and the initial port @anatolykopyl 🤗

**Comment by jbltx** [comment]
Nice feature, do you plan to also support headings from custom JSX/TSX components to appear in ToC ? I have a React component that reads a JSON file and create headings and paragraphs based on it (the JSON file is loaded as a module but I guess it means the ToC should be generated dynamically at runtime and not at compile time?)

**Comment by slorber** [comment]
@jbltx we only support importing other MDX files atm

However, it could be technically possible to also support React imported components, and try to get the compiler to output something looking like this:

```jsx
import Partial, 
  {toc as __tocPartial} from "_partial.mdx"
import MyComponent from "MyComponent"

# Doc

## Doc heading

<Partial/>

<MyComponent/>

export const toc = [
  {
    "value": "Doc heading",
    "id": "doc-heading",
    "level": 2
  }, 
  ...__tocPartial,
  ...MyComponent?.toc?.() 
] 
```


Note that in any case, it would be your responsibility to implement the `toc` function attribute:

```js
const MyComponent = () => <h2 id="hello">Hello</h2>

MyComponent.toc = () => {
  return [{id: "hello",value: "Hello",level: 2}];
};
```

Would that be a helpful addition, or is this too limited?

**Comment by jbltx** [comment]
> Would that be a helpful addition, or is this too limited?

Thank you for the suggestion @slorber, I think your example is enough for my use case indeed! 👍

**Comment by felipecrs** [comment]
I don't know if I'm doing something wrong, but this is still not working for me, even after the fix:

`_release.mdx`:

```mdx
## Release {props.release}
```

`releases.mdx`:
```mdx
import Release from "./_release.mdx"

# Releases

## Top heading

{["v1", "v2", "v3"].map((release) => <Release release={release} />)}
```

And the result:

![image](https://github.com/facebook/docusaurus/assets/29582865/fe77b3f8-d203-4afb-bb7a-3d85e9f0c16e)

**Comment by slorber** [comment]
@felipecrs this is unlikely we will ever support this. The TOC is computed statically at MDX compilation time. We are not running the JS to see what the output is and if it contains any heading. Hence we don't support such loop, nor using props inside partial headings.

If you want that, you'd have to use the TOC function form described just above, a possible enhancement that we haven't implemented yet.

**Comment by felipecrs** [comment]
@slorber I see.

Implementing my own `toc` function is 100% acceptable for my use case, I would appreciate this capability.

**Comment by felipecrs** [comment]
Ops, sorry, did you mean that I could already implement my `toc` function?

I tried it with:

```mdx
import Release from "./_release.mdx"

# Releases

## Top heading

{["v1", "v2", "v3"].map((release) => <Release release={release} />)}

export const toc = () => ["v1", "v2", "v3"].map((release) => ({
  "value": `Release ${release}`,
  "id": `release-${release}`,
  "level": 2
}))
```

And then the toc disappeared entirely.

**Comment by slorber** [comment]
@felipecrs we don't support a toc function yet, only a toc array export.

You should rather do that instead of a loop

```mdx
import ReleaseV1 from "./_releaseV1.mdx"
import ReleaseV2 from "./_releaseV2.mdx"
import ReleaseV3 from "./_releaseV3.mdx"

# Releases

## Top heading

<ReleaseV1/>

<ReleaseV2/>

<ReleaseV3/>
```

Note, we don't support "dynamic headings" such as `## Release {props.version}`, so you will need to provide a TOC export if you want dynamic headings (you can provide the toc export either on the root doc or the partial). There's also a non-implemented proposal to solve that problem that you can track here: https://github.com/facebook/docusaurus/pull/9773

**Comment by felipecrs** [comment]
Great. Thank you, I'll watch the pull request, because my data is very dynamic so I can't define them statically.

**Comment by todo0123** [comment]
Hi guys,

Unfortunately, this issue still isn't fixed, and I can add a note that `<h1>`..`<h6>` tags aren't added either. I know this behavior is described in the docs (that only MD headings go to the ToC), but I have a **proposal**.

**What if the ToC were fully a client-side JS component?**
This right ToC sidebar has nothing to do with either SEO or the main content, so we can safely remove it from static generation.

Then, a JS script added to the page could scan all the headings on the page and build the ToC on the fly!

The ToC would contain headings:

- included from any .tsx or .mdx components
- defined in MD or HTML formats
- inside or outside React code

No complicated server-side parsing, no struggling with preprocessing sequences.

It would be better to allow components to exclude certain H1..H6 elements from the ToC if needed using an attribute like `data-include-to-toc="false"`, but include every existing heading written in all possible forms in the ToC by default.

Any pitfalls? 
What do you think?

**Comment by slorber** [comment]
> Unfortunately, this issue still isn't fixed

The issue as originally described is fixed. If something is not, that must be something slightly different, and you'd rather provide a repro so that we can understand your exact setup.


> I can add a note that `<h1>`..`<h6>` tags aren't added either. 

We only add to TOC Markdown headings, not JSX nodes that produce headings. 

If you think that could be useful, feel free to open a separate proposal. It isn't really related to this issue which is about importing one MDX into another.



> I know this behavior is described in the docs (that only MD headings go to the ToC), but I have a **proposal**.
> 
> **What if the ToC were fully a client-side JS component?** This right ToC sidebar has nothing to do with either SEO or the main content, so we can safely remove it from static generation.
> 
> Then, a JS script added to the page could scan all the headings on the page and build the ToC on the fly!
> 
> The ToC would contain headings:
> 
> * included from any .tsx or .mdx components
> * defined in MD or HTML formats
> * inside or outside React code
> 
> No complicated server-side parsing, no struggling with preprocessing sequences.
> 
> It would be better to allow components to exclude certain H1..H6 elements from the ToC if needed using an attribute like `data-include-to-toc="false"`, but include every existing heading written in all possible forms in the ToC by default.
> 
> Any pitfalls? What do you think?

We purposefully don't want that. 

Having the TOC magically appear after the page loads and React hydrates wouldn't give a very good UX. 

Note that if you believe this tradeoff is acceptable for your site, you can already swizzle the TOC component and build this yourself in a React `useEffect` instead of using the TOC items provided as props. 

Since this is possible to achieve in userland today, I'd suggest that you code this on your site, deploy it to production, and send us the link for the community to see how it behaves.

If there's traction for your proposed solution after seeing it in action, we could eventually design a "toc mode" feature that allows you to opt in for client-side generation. I doubt we'll make it the default mode, though.
